experimental/SetPartitions/src/LinearSetPartition.jl

```diff
@@ -0,0 +1,235 @@
+Return the addition between `p` and `q`.
+
+# Examples
+```jldoctest
+julia> S, d = polynomial_ring(QQ, "d")
+(Univariate polynomial ring in d over QQ, d)
+
+julia> a = linear_partition([(set_partition([1, 2], [1, 1]), S(4)), (set_partition([1, 1], [1, 1]), 4*d)])
+LinearSetPartition{SetPartition, QQPolyRingElem}(Dict{SetPartition, QQPolyRingElem}(SetPartition([1, 2], [1, 1]) => 4, SetPartition([1, 1], [1, 1]) => 4*d))
+
+julia> a + a
+LinearSetPartition{SetPartition, QQPolyRingElem}(Dict{SetPartition, QQPolyRingElem}(SetPartition([1, 2], [1, 1]) => 8, SetPartition([1, 1], [1, 1]) => 8*d))
+```
+"""
+function +(p::LinearSetPartition{S, T}, q::LinearSetPartition{S, T}) where 
+        { S <: AbstractPartition, T <: RingElement } 
+
+    result = deepcopy(p)
+
+    for i in pairs(coefficients(q))
+        coefficients(result)[i[1]] = get(coefficients(result), i[1], 0) + i[2]
+    end
+    
+    return linear_partition(coefficients(result))
+end
+
+"""
+    scale(a::RingElement, p::LinearSetPartition{S, T}) where 
+        { S <: AbstractPartition, T <: RingElement }
+
+Multiply each coefficient in `p` with `a` and return
+the result.
+
+# Examples
+```jldoctest
+julia> S, d = polynomial_ring(QQ, "d")
+(Univariate polynomial ring in d over QQ, d)
+
+julia> scale(S(1) / S(2), linear_partition(Dict(set_partition([1, 2], [1, 1]) => S(8), set_partition([1, 1], [1, 1]) => 8*d)))
+LinearSetPartition{SetPartition, QQPolyRingElem}(Dict{SetPartition, QQPolyRingElem}(SetPartition([1, 2], [1, 1]) => 4, SetPartition([1, 1], [1, 1]) => 4*d))
+```
+"""
+function scale(a::RingElement, p::LinearSetPartition{S, T}) where
+        { S <: AbstractPartition, T <: RingElement }
+    result = Dict{S, T}()
+
+    for (i, n) in pairs(coefficients(p))
+        result[i] = a * n
+    end
+    return linear_partition(result)
+end 
+
+function *(a::RingElement, p::LinearSetPartition)
+    return scale(a, p)
+end
+
+"""
+    compose(p::LinearSetPartition{S, T}, q::LinearSetPartition{S, T}, d::T) where 
+        { S <: AbstractPartition, T <: RingElement }
+
+Multiply each coefficient from `p` with each coefficient from `q`
+as well as perform a composition between each SetPartition part 
+in `p` and `q` and return the result.
+
+# Examples
+```jldoctest
+julia> S, d = polynomial_ring(QQ, "d")
+(Univariate polynomial ring in d over QQ, d)
+
+julia> a = linear_partition([(set_partition([1, 2], [1, 1]), S(4)), (set_partition([1, 1], [1, 1]), 4*d)])
+LinearSetPartition{SetPartition, QQPolyRingElem}(Dict{SetPartition, QQPolyRingElem}(SetPartition([1, 2], [1, 1]) => 4, SetPartition([1, 1], [1, 1]) => 4*d))
+
+julia> compose(a, a, d)
+LinearSetPartition{SetPartition, QQPolyRingElem}(Dict{SetPartition, QQPolyRingElem}(SetPartition([1, 2], [1, 1]) => 16*d + 16, SetPartition([1, 1], [1, 1]) => 16*d^2 + 16*d))
+```
+"""
+function compose(p::LinearSetPartition{S, T}, q::LinearSetPartition{S, T}, d::T) where 
+        { S <: AbstractPartition, T <: RingElement }
+    result = Dict{S, T}()
```

Comment:
Don't you need to check that both linear partitions are indeed defined over the same ring? Eg in the case of number fields you cannot deduce that they have the same ring just from the type parameter T.
(same issue for + and tensorprodukt)